File: cards.cpp
#include "deck.h"
#include <iostream>
#include <unordered_set>

#define SAMPLES 100000L

using namespace std;

int main() {
  int numcards[SAMPLES]; //card count samples
  long sum = 0;
  double average; //average of all samples (the output of the program)
  Deck deck; //deck of cards
  Card card; //drawn card
  unordered_set<string> faces, suits; //keep track of which faces and suits were drawn
  
  for (int i = 0; i < SAMPLES; i++){
    deck.shuffleDeck();
    do{
      card = deck.dealCard();
     
      faces.insert(card.getFace()); //add face to set if not already there
      suits.insert(card.getSuit()); //add suit to set if not already there
        
    }while(faces.size() < 13 || suits.size() < 4); //keep drawing until one of each suit and face has been drawn

    numcards[i] = deck.getCurrentCard(); //how many cards were dealt before getting all suits and faces?

    faces.clear(); //reset for next sample
    suits.clear(); //reset for next sample
  }

  for (int i = 0; i < SAMPLES; i++){
    sum += numcards[i];
  }
  average = sum / (double) SAMPLES;
  cout << "The average number of cards dealt is " << average << endl;
  
  return 0;
} // main()

File: cards.py
import random

def shuffle_deck():
    ranks = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']
    suits = ['Hearts', 'Diamonds', 'Clubs', 'Spades']
    deck = [{'rank': rank, 'suit': suit} for rank in ranks for suit in suits]
    random.shuffle(deck)
    return deck

def draw_cards(deck):
    drawn_cards = set()
    required_ranks = set(['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'])
    required_suits = set(['Hearts', 'Diamonds', 'Clubs', 'Spades'])
    drawn_ranks = set()
    drawn_suits = set()
    num_drawn_cards = 0

    while drawn_ranks != required_ranks or drawn_suits != required_suits:
        card = deck.pop()
        print(card) #lrb
        num_drawn_cards += 1
        drawn_cards.add((card['rank'], card['suit']))
        drawn_ranks.add(card['rank'])
        drawn_suits.add(card['suit'])

    return num_drawn_cards, drawn_cards

# Shuffle the deck
shuffled_deck = shuffle_deck()

# Draw cards until all ranks and suits are present
num_drawn_cards, drawn_cards = draw_cards(shuffled_deck)

# Print the result
print("Number of cards drawn:", num_drawn_cards)

